s3upload/miniwdl_s3upload.py

```diff
@@ -26,16 +26,110 @@
 """
 
 import os
+from os.path import relpath
+from os import path
 import subprocess
 import threading
 import json
+import logging
+from pathlib import Path
+from urllib.parse import urlparse
+from typing import Dict, Tuple
+
 import WDL
+from WDL import Env, Value, values_to_json
+from WDL.runtime import cache, config
 from WDL._util import StructuredLogMessage as _
 
-_uploaded_files = {}
+import boto3
+
+s3 = boto3.resource("s3", endpoint_url=os.getenv("AWS_ENDPOINT_URL"))
+s3_client = boto3.client("s3", endpoint_url=os.getenv("AWS_ENDPOINT_URL"))
+
+
+def s3_object(uri: str):
+    assert uri.startswith("s3://")
+    bucket, key = uri.split("/", 3)[2:]
+    return s3.Bucket(bucket).Object(key)
+
+
+def get_s3prefix(cfg: config.Loader) -> str:
+    s3prefix = cfg["s3_progressive_upload"]["uri_prefix"]
+    assert s3prefix.startswith("s3://"), "MINIWDL__S3_PROGRESSIVE_UPLOAD__URI_PREFIX invalid"
+    return s3prefix
+
+
+def s3uri(cfg: config.Loader, key: str) -> str:
+    return os.path.join(get_s3prefix(cfg), "cache", f"{key}.json")
+
+
+def flag_temporary(s3uri):
+    uri = urlparse(s3uri)
+    bucket, key = uri.hostname, uri.path[1:]
+    s3_client.put_object_tagging(
+        Bucket=bucket,
+        Key=key,
+        Tagging={
+            'TagSet': [
+                {
+                    'Key': 'swipe_temporary',
+                    'Value': 'true'
+                },
+            ]
+        },
+    )
+
+
+def inode(link: str):
+    st = os.stat(os.path.realpath(link))
+    return (st.st_dev, st.st_ino)
+
+
+_uploaded_files: Dict[Tuple[int, int], str] = {}
+_cached_files: Dict[Tuple[int, int], Tuple[str, Env.Bindings[Value.Base]]] = {}
 _uploaded_files_lock = threading.Lock()
 
 
+def cache_put(cfg: config.Loader, logger: logging.Logger, key: str, outputs: Env.Bindings[Value.Base]):
+    if not (cfg["call_cache"].get_bool("put") and
+            cfg["call_cache"]["backend"] == "s3_progressive_upload_call_cache_backend"):
+        return
+
+    json_values = values_to_json(outputs)
+    file_outputs: Dict[str, str] = {k: v for k, v in json_values.items() if path.exists(v)}
```

Comment:
This only supports *file* outputs, right? Some of our WDL's use *string* inputs and outputs. Is the recommendation to only use files if we want to enable call caching?